content/blogposts/2024/anatomy-of-a-coding-assistant.md

```diff
@@ -0,0 +1,166 @@
+    1. For free and Pro users, we use our home-grown search engine called symf, written in Golang, that runs on your local environment.
+    2. For Enterprise users, we leverage [Sourcegraph’s Code Search](https://sourcegraph.com/code-search) for this.
+    
+![10_indexed](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/10_indexed.png)
+    
+2. **Embeddings:** When we set up embeddings, we first convert code snippets from various files into embeddings in a preprocessing step. Then, when we get the user query, we convert it to an embedding as well. Using these embeddings, we perform vector search to find the most relevant code snippets for the user query. An embedding is a vector representation of a piece of text, and vector search helps find the most relevant pieces of text given another piece of text, such as the user query.
+    1. To convert code snippets into embeddings, we use either OpenAI’s text-embedding-ada-002 model or Sourcegraph's own model called st-multi-qa-mpnet-base-dot-v1. The choice between these models is [determined by a feature flag](https://github.com/sourcegraph/cody/blob/ee47055967bdb3e65c8694cab726dd47d50c29ba/vscode/src/local-context/local-embeddings.ts#L31). The embeddings are [stored locally on your machine](https://github.com/sourcegraph/cody/blob/main/vscode/src/local-context/local-embeddings.ts#L55).
+    
+![11_embeddings](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/11_embeddings.png)
+    
+
+In our product, [we default to a mix of both](https://github.com/sourcegraph/cody/blob/0238df0eda7e6e9be03139a56a89e235b76be448/vscode/package.json#L967) with an option to use one or the other.
+
+### Recent improvements to context fetching for chat
+
+With [the June release](https://sourcegraph.com/blog/release/june-2024), we made several improvements: Enterprise context now rewrites queries, enabling us to handle longer and foreign-language questions. We also detect filenames and symbols in queries, addressing a class of "easy-looking" questions users expect us to get right.
+
+---
+
+## Context sources outside of your codebase
+
+In addition to automatic context fetching, we've found that providing a way for the user to manually include context is also helpful. We've done this by letting the user ask questions about specific files by @-mentioning them. More recently, we've realized that there is more context the user sometimes wants to include outside of the codebase, such as documentation, web pages, Slack threads, Linear issues, and Notion documents. We've made this possible through [OpenCtx](https://openctx.org/), and it has proven useful.
+
+[![20_tweet](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/20_tweet.png)](https://x.com/beyang/status/1792616243742244974)
+
+At the moment, this allows users to include context from various sources manually. In the future, we envision this feature being used by both users and AI.
```

Comment:
```suggestion
```

It's accessible to both humans and AI today. OpenCtx annotations are visible in the Sourcegraph UI.